components/common/src/lib.rs

```diff
@@ -146,7 +146,171 @@ impl FeatureFlag {
 }
 
 pub mod sync {
-    use std::time::Duration;
+    use habitat_core::env::Config as EnvConfig;
+    use std::{collections::HashMap,
+              sync::Mutex,
+              thread::{self,
+                       ThreadId},
+              time::{Duration,
+                     Instant}};
+
+    type NameAndLastHeartbeat = (Option<String>, Instant);
+    type HeartbeatMap = HashMap<ThreadId, NameAndLastHeartbeat>;
+    lazy_static::lazy_static! {
+        static ref THREAD_HEARTBEATS: Mutex<HeartbeatMap> = Default::default();
+    }
+
+    struct ThreadAliveThreshold(Duration);
+
+    impl EnvConfig for ThreadAliveThreshold {
+        const ENVVAR: &'static str = "HAB_THREAD_ALIVE_THRESHOLD_SECS";
+    }
+
+    impl Default for ThreadAliveThreshold {
+        fn default() -> Self { Self(Duration::from_secs(5 * 60)) }
+    }
+
+    impl std::str::FromStr for ThreadAliveThreshold {
+        type Err = std::num::ParseIntError;
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            Ok(Self(Duration::from_secs(s.parse()?)))
+        }
+    }
+
+    struct ThreadAliveCheckDelay(Duration);
+
+    impl EnvConfig for ThreadAliveCheckDelay {
+        const ENVVAR: &'static str = "HAB_THREAD_ALIVE_THRESHOLD_SECS";
+    }
+
+    impl Default for ThreadAliveCheckDelay {
+        fn default() -> Self { Self(Duration::from_secs(60)) }
+    }
+
+    impl std::str::FromStr for ThreadAliveCheckDelay {
+        type Err = std::num::ParseIntError;
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            Ok(Self(Duration::from_secs(s.parse()?)))
+        }
+    }
+
+    /// Call periodically from a thread which has a work loop to indicate that the thread is
+    /// still alive and processing its loop. If this function is not called for more than
+    /// `ThreadAliveThreshold`, it will be error logged as a likely deadlock.
+    pub fn mark_thread_alive() {
+        mark_thread_alive_impl(&mut THREAD_HEARTBEATS.lock()
+                                                     .expect("THREAD_HEARTBEATS poisoned"));
+    }
+
+    fn mark_thread_alive_impl(heartbeats: &mut HeartbeatMap) {
+        let thread = thread::current();
+        heartbeats.insert(thread.id(),
+                          (thread.name().map(str::to_string), Instant::now()));
+    }
+
+    /// Call once per binary to start the thread which will check that all the threads that
+    /// call `mark_thread_alive` continue to do so.
+    pub fn spawn_thread_alive_checker() {
+        thread::Builder::new().name("thread-alive-check".to_string())
+                              .spawn(|| {
+                                  loop {
+                                      check_thread_heartbeats();
+                                      thread::sleep(ThreadAliveCheckDelay::configured_value().0);
```

Comment:
Similarly, can we figure out what our delay is once before entering the loop since it will never change?